### PR TITLE
Add single point calculations

### LINKED ISSUE #30: create single point calculation driver
- [ ] integrate mlip multicalculator 
- [ ] create single point calculators
- [ ] create tests
- [ ] create documentation

### PR DESCRIPTION
**Pull request by ElliottKasoar**
Resolves #30

Initial code to configure calculators and perform single point calculations for energies, forces and stress.

Largely based on code already in [lavello](https://gitlab.com/drFaustroll/lavello/), but refactors, fixes various issues to get tests etc. to pass, and adds more general tests.

Now also adds the ability to return forces and stress.

To do:

- [x] Fix pylint
  - I had previously fixed with with try...except clauses, but it makes it quite verbose. Disabling pylint's complaint is an alternative, since the error message is pretty understandable.
- [x]  Add support for trajectory energies
  - The obvious solution is change `read` to read a specified slice/all structures, and attach the calculator to each image.
  - Are there better ways of doing this? As far as I can tell, calculators can only be attached to `Atoms` objects.
- [x] Update docstrings in various places
- [x] Add single point forces, stresses etc.
  - How (probably a dictionary)/when do we want to return these? 
- [x] Replace UiO66 with smaller structure that we are still able to calculate stress for e.g. NaCl from @alinelena

Note: CI is currently not working due to issues with pulling ASE from GitLab, but all tests pass locally.

### CODE REVIEW

File: janus_core/single_point.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            device=self.device,
+            **kwargs,
+        )
+        if self.sys is None:
+            self.read_system()
+        self.sys.calc = calculator
+
+    def get_potential_energy(self) -> float:
+        """Calculate potential energy using MLIP."""
+        return self.sys.get_potential_energy()
```
Comment (alinelena) [inline janus_core/single_point.py:61]: ok we shall return actually for single point more than just the energy, forces, stresses, etc...

**Comment by ElliottKasoar** [comment]
Current temporary solution to ASE installation is to update the commit once tests start failing, but this may be a problem for publishing etc. if it passes here, but fails when rerun.

File: tests/test_single_point.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    results = single_point.run_single_point()
+    for prop in ["energy", "forces", "stress"]:
+        assert prop in results
+
+
+def test_single_point_traj():
+    """Test single point stress using MACE calculator."""
+    data_path = os.path.join(os.path.dirname(__file__), "data", "benzene-traj.xyz")
+    model_path = os.path.join(os.path.dirname(__file__), "models", "MACE_small.model")
```
Comment (alinelena) [inline tests/test_single_point.py:71]: mace_small shall be mace_mp_small

File: tests/test_single_point.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
+"""Test configuration of MLIP calculators."""
+
+import os
```
Comment (alinelena) [inline tests/test_single_point.py:3]: use pathlib everywhere

File: janus_core/mlip_calculators.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+Similar in spirit with matcalc and quacc approaches
+- https://github.com/materialsvirtuallab/matcalc
+- https://github.com/Quantum-Accelerators/quacc.git
+"""
+
+from __future__ import annotations
+
+from typing import Literal
+
+from ase.calculators.calculator import Calculator
```
Comment (alinelena) [inline janus_core/mlip_calculators.py:12]: check if we can get away with latest released ase, and if yes, put a note in README for the users for serious matters they shall install ase from git.

**Review by alinelena — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -23,6 +23,8 @@ pre-commit install  # install pre-commit hooks
 pytest -v  # discover and run all tests
 ```
 
+Manually updating ASE via https://gitlab.com/ase/ase is strongly recommended, as tags are no longer regularly published.
```
Comment (alinelena) [inline README.md:26]: you can put also pip install git+https.... command

**Review by alinelena — APPROVED** [review]

File: janus_core/mlip_calculators.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    # pylint: disable=too-many-branches
+    # pylint: disable=import-error
+    # Optional imports handled via `architecture`. We could catch these,
+    # but the error message is clear if imports are missing.
+    if architecture == "mace":
+        from mace import __version__
+        from mace.calculators import MACECalculator
+
+        if "default_dtype" not in kwargs:
+            kwargs["default_dtype"] = "float64"
```
Comment (oerc0122) [inline janus_core/mlip_calculators.py:49]: [`dict.setdefault`](https://docs.python.org/3/library/stdtypes.html#dict.setdefault) is there to handle this type of thing
```
kwargs.setdefault("default_dtype", "float64")
```

File: janus_core/mlip_calculators.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        Invalid architecture specified.
+
+    Returns
+    -------
+    calculator : Calculator
+        Configured MLIP calculator.
+    """
+    # pylint: disable=import-outside-toplevel
+    # pylint: disable=too-many-branches
+    # pylint: disable=import-error
```
Comment (oerc0122) [inline janus_core/mlip_calculators.py:41]: Can comma separate:
```
pylint: disable=import-outside-toplevel, too-many-branches, import-error
```

File: janus_core/mlip_calculators.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        calculator = mace_off(**kwargs)
+
+    elif architecture == "m3gnet":
+        from matgl import __version__, load_model
+        from matgl.ext.ase import M3GNetCalculator
+
+        if "model" not in kwargs:
+            model = load_model("M3GNet-MP-2021.2.8-DIRECT-PES")
+        if "stress_weight" not in kwargs:
+            kwargs.setdefault("stress_weight", 1.0 / 160.21766208)
```
Comment (oerc0122) [inline janus_core/mlip_calculators.py:85]: The poinnt of `setdefault` is that you don't need the if.

File: janus_core/single_point.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+
+class SinglePoint:
+    """Perpare and perform single point calculations."""
+
+    #  pylint: disable=dangerous-default-value
+    def __init__(
+        self,
+        system: str,
+        architecture: str = "mace_mp",
```
Comment (oerc0122) [inline janus_core/single_point.py:20]: Can use `architectures` in `mlip_calculators` here.

File: janus_core/single_point.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        Returns
+        -------
+        potential_energy : float | list[float]
+            Potential energy of system(s).
+        """
+        if isinstance(self.sys, list):
+            energies = []
+            for sys in self.sys:
+                energies.append(sys.get_potential_energy())
+            return energies
```
Comment (oerc0122) [inline janus_core/single_point.py:91]:
```
            energies = [sys.get_potential_energy() for sys in self.sys]
            return energies
```

File: janus_core/single_point.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+        Returns
+        -------
+        forces : ndarray | list[ndarray]
+            Forces of system(s).
+        """
+        if isinstance(self.sys, list):
+            forces = []
+            for sys in self.sys:
+                forces.append(sys.get_forces())
```
Comment (oerc0122) [inline janus_core/single_point.py:106]: As above

File: janus_core/single_point.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+        Returns
+        -------
+        stress : ndarray | list[ndarray]
+            Stress of system(s).
+        """
+        if isinstance(self.sys, list):
+            stress = []
+            for sys in self.sys:
+                stress.append(sys.get_stress())
```
Comment (oerc0122) [inline janus_core/single_point.py:122]: As above

File: tests/test_mlip_calculators.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+from janus_core.mlip_calculators import choose_calculator
+
+
+def test_mace():
+    """Test mace calculator can be configured."""
+    model_path = Path(Path(__file__).parent, "models", "mace_mp_small.model")
+    calculator = choose_calculator(
+        architecture="mace", device="cpu", model_paths=model_path
+    )
+    assert calculator.parameters["version"] is not None
```
Comment (oerc0122) [inline tests/test_mlip_calculators.py:16]: These could be `pytest.mark.parametrize`

File: tests/test_single_point.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+from pathlib import Path
+
+import pytest
+
+from janus_core.single_point import SinglePoint
+
+
+def test_potential_energy():
+    """Test single point energy using MACE calculator."""
+    data_path = Path(Path(__file__).parent, "data", "benzene.xyz")
```
Comment (oerc0122) [inline tests/test_single_point.py:12]: The `/` is the path join operator

**Review by oerc0122 — COMMENTED** [review]
Not sure if these will come through.

**Pull request merged**